domain-models-runtime/src/main/java/org/folio/rest/persist/Conn.java

```diff
@@ -297,4 +306,133 @@ static String durationMsg(String descriptionKey, String sql, long startNanoTime)
+    return new LocalRowSet(0).withColumns(Collections.singletonList("id"));
+  }
+
+  /**
+   * @throws Throwable the caller needs to catch and convert into a failed Future
+   */
+  private Future<RowSet<Row>> saveBatchInternal(boolean upsert, String table, List<Tuple> batch) {
+
+    if (batch.isEmpty()) {
+      // vertx-pg-client fails with "Can not execute batch query with 0 sets of batch parameters."
+      return Future.succeededFuture(emptyRowSetOfId());
+    }
+    long start = log.isDebugEnabled() ? System.nanoTime() : 0;
+    log.info("starting: saveBatch size=" + batch.size());
+    String sql = "INSERT INTO " + postgresClient.getSchemaName() + "." + table
+        + " (id, jsonb) VALUES ($1, $2)"
+        + (upsert ? " ON CONFLICT (id) DO UPDATE SET jsonb = EXCLUDED.jsonb" : "")
+        + " RETURNING id";
+    return pgConnection.preparedQuery(sql).executeBatch(batch)
+    .map(rowSet -> {
+      log.debug(() -> durationMsg("saveBatch", table, start));
+      if (rowSet == null) {
+        return emptyRowSetOfId();
+      }
+      return rowSet;
+    })
+    .onFailure(e -> {
+      log.error("saveBatch size=" + batch.size() + " " + e.getMessage(), e);
+      log.debug(() -> durationMsg("saveBatchFailed", table, start));
+    });
+  }
+
+  /**
+   * Insert the entities into table using a single INSERT statement.
+   * @param table  destination table to insert into
+   * @param entities  each array element is a String with the content for the JSONB field of table; if id is missing a random id is generated
+   * @return one result row per inserted row, containing the id field
+   */
+  public Future<RowSet<Row>> saveBatch(String table, JsonArray entities) {
+    return saveBatch(false, table, entities);
+  }
+
+  /**
+   * Upsert the entities into table using a single INSERT statement.
+   * @param table  destination table to insert into
+   * @param entities  each array element is a String with the content for the JSONB field of table; if id is missing a random id is generated
+   * @return one result row per inserted row, containing the id field
+   */
+  public Future<RowSet<Row>> upsertBatch(String table, JsonArray entities) {
+    return saveBatch(true, table, entities);
+  }
+
+  <T> Future<RowSet<Row>> saveBatch(boolean upsert, String table, List<T> entities) {
+
+    try {
+      List<Tuple> batch = new ArrayList<>();
+      if (entities == null || entities.isEmpty()) {
+        return Future.succeededFuture(emptyRowSetOfId());
+      }
+      // We must use reflection, the POJOs don't have a interface/superclass in common.
+      Method getIdMethod = entities.get(0).getClass().getDeclaredMethod("getId");
+      for (Object entity : entities) {
+        Object obj = getIdMethod.invoke(entity);
+        UUID id = obj == null ? UUID.randomUUID() : UUID.fromString((String) obj);
+        batch.add(Tuple.of(id, PostgresClient.pojo2JsonObject(entity)));
+      }
+      return saveBatchInternal(upsert, table, batch);
```

Comment:
Would it work (and could be simpler) if you convert `List<T>` to `JsonArray` and call `saveBatch(boolean upsert, String table, JsonArray entities)` directly?

Reply:
`Conn#saveBatch(boolean upsert, String table, JsonArray entities)` calls
`Conn#saveBatchInternal(boolean upsert, String table, List<Tuple> batch)`.

`Conn#saveBatchInternal(boolean upsert, String table, List<Tuple> batch)` calls
`[PreparedQuery#executeBatch(List<Tuple> batch)](https://vertx.io/docs/apidocs/io/vertx/sqlclient/PreparedQuery.html#executeBatch-java.util.List-)`.

We should directly convert to `List<Tuple>`. I don't see an advantage in converting to `JsonArray` that requires another conversion to `List<Tuple>`.

Reply:
My point is why do you need the reflection call? If you can assume there is a `getId` method you can also convert the object to json directly and assume there is an `id` field.

Reply:
The code has been moved from https://github.com/folio-org/raml-module-builder/blob/v32.2.0/domain-models-runtime/src/main/java/org/folio/rest/persist/PostgresClient.java#L1181 to Conn.java.
PostgresClient.pojo2JsonObject uses Jackson ObjectMapper that probably also uses reflection.
The change you suggest requires careful investigation and is out of scope of RMB-602. Please create a new Jira for it.